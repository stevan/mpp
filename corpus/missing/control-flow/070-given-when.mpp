# Basic given/when structure
given ($value) {
    when (1) {
        print("one");
    }
    when (2) {
        print("two");
    }
    when (3) {
        print("three");
    }
    default {
        print("other");
    }
}

# Given with variable
given ($x) {
    when ($_ < 0) {
        print("negative");
    }
    when ($_ > 0) {
        print("positive");
    }
    default {
        print("zero");
    }
}

# Given with expression
given ($x + $y) {
    when (10) {
        process_ten();
    }
    when (20) {
        process_twenty();
    }
    when (30) {
        process_thirty();
    }
}

# Nested given/when
given ($category) {
    when ("number") {
        given ($value) {
            when (0) {
                print("zero");
            }
            when (1) {
                print("one");
            }
            default {
                print("other number");
            }
        }
    }
    when ("string") {
        print("string value");
    }
}

# When with multiple values
given ($day) {
    when ([1, 2, 3, 4, 5]) {
        print("weekday");
    }
    when ([6, 7]) {
        print("weekend");
    }
}

# When with regex (without smartmatch operator)
given ($input) {
    when (/^[0-9]+$/) {
        print("numeric");
    }
    when (/^[a-z]+$/) {
        print("alphabetic");
    }
    default {
        print("mixed");
    }
}

# Given without default
given ($code) {
    when (200) {
        success();
    }
    when (404) {
        not_found();
    }
    when (500) {
        server_error();
    }
}

# Empty when blocks
given ($flag) {
    when (0) {
    }
    when (1) {
    }
}

# Break statement in when
given ($action) {
    when ("stop") {
        cleanup();
        break;
    }
    when ("pause") {
        save_state();
        break;
    }
}

# Continue to next when
given ($score) {
    when ($_ > 90) {
        print("A");
        continue;
    }
    when ($_ > 80) {
        print("B");
        continue;
    }
    when ($_ > 70) {
        print("C");
    }
}