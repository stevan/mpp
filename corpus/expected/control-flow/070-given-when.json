[
  {
    "type": "Given",
    "expression": {
      "type": "Variable",
      "name": "$value"
    },
    "whenClauses": [
      {
        "condition": {
          "type": "Number",
          "value": "1"
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"one\""
              }
            ]
          }
        ]
      },
      {
        "condition": {
          "type": "Number",
          "value": "2"
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"two\""
              }
            ]
          }
        ]
      },
      {
        "condition": {
          "type": "Number",
          "value": "3"
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"three\""
              }
            ]
          }
        ]
      }
    ],
    "defaultBlock": [
      {
        "type": "Call",
        "name": "print",
        "arguments": [
          {
            "type": "String",
            "value": "\"other\""
          }
        ]
      }
    ]
  },
  {
    "type": "Given",
    "expression": {
      "type": "Variable",
      "name": "$x"
    },
    "whenClauses": [
      {
        "condition": {
          "type": "BinaryOp",
          "operator": "<",
          "left": {
            "type": "Variable",
            "name": "$_"
          },
          "right": {
            "type": "Number",
            "value": "0"
          }
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"negative\""
              }
            ]
          }
        ]
      },
      {
        "condition": {
          "type": "BinaryOp",
          "operator": ">",
          "left": {
            "type": "Variable",
            "name": "$_"
          },
          "right": {
            "type": "Number",
            "value": "0"
          }
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"positive\""
              }
            ]
          }
        ]
      }
    ],
    "defaultBlock": [
      {
        "type": "Call",
        "name": "print",
        "arguments": [
          {
            "type": "String",
            "value": "\"zero\""
          }
        ]
      }
    ]
  },
  {
    "type": "Given",
    "expression": {
      "type": "BinaryOp",
      "operator": "+",
      "left": {
        "type": "Variable",
        "name": "$x"
      },
      "right": {
        "type": "Variable",
        "name": "$y"
      }
    },
    "whenClauses": [
      {
        "condition": {
          "type": "Number",
          "value": "10"
        },
        "block": [
          {
            "type": "Call",
            "name": "process_ten",
            "arguments": []
          }
        ]
      },
      {
        "condition": {
          "type": "Number",
          "value": "20"
        },
        "block": [
          {
            "type": "Call",
            "name": "process_twenty",
            "arguments": []
          }
        ]
      },
      {
        "condition": {
          "type": "Number",
          "value": "30"
        },
        "block": [
          {
            "type": "Call",
            "name": "process_thirty",
            "arguments": []
          }
        ]
      }
    ]
  },
  {
    "type": "Given",
    "expression": {
      "type": "Variable",
      "name": "$category"
    },
    "whenClauses": [
      {
        "condition": {
          "type": "String",
          "value": "\"number\""
        },
        "block": [
          {
            "type": "Given",
            "expression": {
              "type": "Variable",
              "name": "$value"
            },
            "whenClauses": [
              {
                "condition": {
                  "type": "Number",
                  "value": "0"
                },
                "block": [
                  {
                    "type": "Call",
                    "name": "print",
                    "arguments": [
                      {
                        "type": "String",
                        "value": "\"zero\""
                      }
                    ]
                  }
                ]
              },
              {
                "condition": {
                  "type": "Number",
                  "value": "1"
                },
                "block": [
                  {
                    "type": "Call",
                    "name": "print",
                    "arguments": [
                      {
                        "type": "String",
                        "value": "\"one\""
                      }
                    ]
                  }
                ]
              }
            ],
            "defaultBlock": [
              {
                "type": "Call",
                "name": "print",
                "arguments": [
                  {
                    "type": "String",
                    "value": "\"other number\""
                  }
                ]
              }
            ]
          }
        ]
      },
      {
        "condition": {
          "type": "String",
          "value": "\"string\""
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"string value\""
              }
            ]
          }
        ]
      }
    ]
  },
  {
    "type": "Given",
    "expression": {
      "type": "Variable",
      "name": "$day"
    },
    "whenClauses": [
      {
        "condition": {
          "type": "ArrayLiteral",
          "elements": [
            {
              "type": "Number",
              "value": "1"
            },
            {
              "type": "Number",
              "value": "2"
            },
            {
              "type": "Number",
              "value": "3"
            },
            {
              "type": "Number",
              "value": "4"
            },
            {
              "type": "Number",
              "value": "5"
            }
          ]
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"weekday\""
              }
            ]
          }
        ]
      },
      {
        "condition": {
          "type": "ArrayLiteral",
          "elements": [
            {
              "type": "Number",
              "value": "6"
            },
            {
              "type": "Number",
              "value": "7"
            }
          ]
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"weekend\""
              }
            ]
          }
        ]
      }
    ]
  },
  {
    "type": "Given",
    "expression": {
      "type": "Variable",
      "name": "$input"
    },
    "whenClauses": [
      {
        "condition": {
          "type": "RegexLiteral",
          "pattern": "^[0-9]+$",
          "flags": ""
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"numeric\""
              }
            ]
          }
        ]
      },
      {
        "condition": {
          "type": "RegexLiteral",
          "pattern": "^[a-z]+$",
          "flags": ""
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"alphabetic\""
              }
            ]
          }
        ]
      }
    ],
    "defaultBlock": [
      {
        "type": "Call",
        "name": "print",
        "arguments": [
          {
            "type": "String",
            "value": "\"mixed\""
          }
        ]
      }
    ]
  },
  {
    "type": "Given",
    "expression": {
      "type": "Variable",
      "name": "$code"
    },
    "whenClauses": [
      {
        "condition": {
          "type": "Number",
          "value": "200"
        },
        "block": [
          {
            "type": "Call",
            "name": "success",
            "arguments": []
          }
        ]
      },
      {
        "condition": {
          "type": "Number",
          "value": "404"
        },
        "block": [
          {
            "type": "Call",
            "name": "not_found",
            "arguments": []
          }
        ]
      },
      {
        "condition": {
          "type": "Number",
          "value": "500"
        },
        "block": [
          {
            "type": "Call",
            "name": "server_error",
            "arguments": []
          }
        ]
      }
    ]
  },
  {
    "type": "Given",
    "expression": {
      "type": "Variable",
      "name": "$flag"
    },
    "whenClauses": [
      {
        "condition": {
          "type": "Number",
          "value": "0"
        },
        "block": []
      },
      {
        "condition": {
          "type": "Number",
          "value": "1"
        },
        "block": []
      }
    ]
  },
  {
    "type": "Given",
    "expression": {
      "type": "Variable",
      "name": "$action"
    },
    "whenClauses": [
      {
        "condition": {
          "type": "String",
          "value": "\"stop\""
        },
        "block": [
          {
            "type": "Call",
            "name": "cleanup",
            "arguments": []
          },
          {
            "type": "Break"
          }
        ]
      },
      {
        "condition": {
          "type": "String",
          "value": "\"pause\""
        },
        "block": [
          {
            "type": "Call",
            "name": "save_state",
            "arguments": []
          },
          {
            "type": "Break"
          }
        ]
      }
    ]
  },
  {
    "type": "Given",
    "expression": {
      "type": "Variable",
      "name": "$score"
    },
    "whenClauses": [
      {
        "condition": {
          "type": "BinaryOp",
          "operator": ">",
          "left": {
            "type": "Variable",
            "name": "$_"
          },
          "right": {
            "type": "Number",
            "value": "90"
          }
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"A\""
              }
            ]
          },
          {
            "type": "Continue"
          }
        ]
      },
      {
        "condition": {
          "type": "BinaryOp",
          "operator": ">",
          "left": {
            "type": "Variable",
            "name": "$_"
          },
          "right": {
            "type": "Number",
            "value": "80"
          }
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"B\""
              }
            ]
          },
          {
            "type": "Continue"
          }
        ]
      },
      {
        "condition": {
          "type": "BinaryOp",
          "operator": ">",
          "left": {
            "type": "Variable",
            "name": "$_"
          },
          "right": {
            "type": "Number",
            "value": "70"
          }
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"C\""
              }
            ]
          }
        ]
      }
    ]
  }
]